core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.plan.hep.HepRelVertex;
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Union;
+import org.apache.calcite.rel.core.Values;
+import org.apache.calcite.rex.RexLiteral;
+
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import org.immutables.value.Value;
+
+import java.util.List;
+
+/**
+ * Planner rule that convert
+ * {@link org.apache.calcite.rel.core.Union} with
+ * input only containing {@link org.apache.calcite.rel.core.Values}
+ * into an {@link org.apache.calcite.rel.core.Values}.
+ *
+ * <p>For example,
+ *
+ * <blockquote><pre>{@code
+ * Union(all=[true])
+ *     Values(tuples=[[{ 3, null }]])
+ *     Values(tuples=[[{ 7369, null }]])
+ *     Values(tuples=[[{ 1, 2 }]])
+ * }</pre></blockquote>
+ *
+ * <p>becomes
+ *
+ * <blockquote><pre>{@code
+ * Values(tuples=[[{ 3, null }, { 7369, null }, { 1, 2 }]])
+ * }</pre></blockquote>
+ *
+ * <p>If (<code>all</code> = <code>false</code>),
+ * {@link org.apache.calcite.rel.core.Values} value will be set list.
+ *
+ * <p>If {@link org.apache.calcite.rel.core.Union} only one input, will be handled by
+ * {@link org.apache.calcite.rel.rules.CoreRules#UNION_REMOVE}
+ *
+ */
+@Value.Enclosing
+public class UnionToValuesRule extends RelRule<UnionToValuesRule.Config>
+    implements TransformationRule {
+
+  /**
+   * Creates a UnionToValuesRule.
+   */
+  protected UnionToValuesRule(UnionToValuesRule.Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public boolean matches(RelOptRuleCall call) {
+    Union union = call.rel(0);
+    return union.getInputs().size() > 1;
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Union union = call.rel(0);
+    List<RelNode> relNodeList = union.getInputs();
+    if (!relNodeList.stream().allMatch(UnionToValuesRule::isValues)) {
+      return;
+    }
+    ImmutableCollection.Builder<ImmutableList<RexLiteral>> tupleList =
+        getImmutableListBuilder(union.all, relNodeList);
+    Values values =
+        (Values) call.builder().values(tupleList.build(), union.getRowType()).build();
+    call.transformTo(values);
+  }
+
+  private static ImmutableCollection.Builder<ImmutableList<RexLiteral>> getImmutableListBuilder(
+      Boolean all, List<RelNode> relNodeList) {
+    ImmutableCollection.Builder<ImmutableList<RexLiteral>> tupleList;
+    if (all) {
+      tupleList = ImmutableList.builder();
+    } else {
+      tupleList = ImmutableSet.builder();
```

Comment:
What is concerning here is that two VALUES may have the same "value" but not be "equal".
In particular, I am not sure about implicit casts.
Perhaps you can add some tests like `VALUE(5.0) UNION ALL VALUES(5)` or
`VALUES(CAST 5.0 AS INT) UNION ALL VALUES(5)`

Reply:
Maybe we will find another issue with `VALUE(5.0) UNION ALL VALUES(5)`. 
Here is a test case.
```
!use post
select *
from (values (5.0)) as t(x)
union
select *
from (values (5)) as t(y);
X
5
5.0
!ok

EnumerableUnion(all=[false])
  EnumerableValues(tuples=[[{ 5.0 }]])
  EnumerableValues(tuples=[[{ 5 }]])
!plan

select *
from (values (cast(5.0 as int))) as t(x)
union
select *
from (values (5)) as t(y);
X
5
!ok

EnumerableUnion(all=[false])
  EnumerableValues(tuples=[[{ 5 }]])
  EnumerableValues(tuples=[[{ 5 }]])
!plan
```
But in Postgresql. The SQL will return 5.0:
```
select *
from (values (5.0)) as t(x)
union
select *
from (values (5)) as t(y);
```

Reply:
Calcite's pervasive use of implicit casts makes solving such problems much harder than expected. Other compilers have a stage which converts implicit casts into explicit casts in the program representation, which is then always unambiguously typed. I think SqlToRel should do that, but that seems like a very difficult task.

Reply:
I agree with your opinion. In Calcite, implicit cast transformations are in different stages according to different operations. In this PR, I have kept this result `5.0 5`. We can move to another issue to resolve this. Now I'm not sure how big the impact of the issue is.

Reply:
The problem is that your transformation will no longer produce an equivalent program. If the UNION removes duplicates, and after transformation you have just VALUES with duplicates, the two programs are no longer computing the same result.

Reply:
you could check to see if the type of all VALUES is the same, and then you can still apply the transformation as it is, otherwise you leave the UNION in the program. The transformation will probably be correct for UNION ALL.

Reply:
I added type validation according to your suggestion, and it seems that the result is correct.